presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -346,29 +354,27 @@ public void moreMemoryAvailable()
 
     public void transferMemoryToTaskContext(long taskBytes)
     {
-        long bytes = memoryReservation.getAndSet(0);
-        driverContext.transferMemoryToTaskContext(bytes);
-
-        TaskContext taskContext = driverContext.getPipelineContext().getTaskContext();
-        if (taskBytes > bytes) {
-            try {
-                taskContext.reserveMemory(taskBytes - bytes);
-            }
-            catch (ExceededMemoryLimitException e) {
-                taskContext.freeMemory(bytes);
-                throw e;
-            }
+        long bytes = operatorMemoryContext.getUserMemory();
+        LocalMemoryContext taskLocalMemoryContext = driverContext.getPipelineContext().getTaskContext().localUserMemoryContext();
+        try {
+            operatorMemoryContext.localUserMemoryContext().transferOwnership(taskLocalMemoryContext);
+            taskLocalMemoryContext.addBytes(taskBytes - bytes);
         }
-        else {
-            taskContext.freeMemory(bytes - taskBytes);
+        catch (ExceededMemoryLimitException e) {
+            // Since we can reserve an extra amount of memory above (when taskBytes > bytes) ExceededMemoryLimitException can be thrown
+            // when reserving this extra amount. This will happen after "bytes" is moved from the operator context to the task context.
+            // So, at this point "bytes" won't be tracked by the operator context and when the driver is destroyed it won't
+            // be freed up from the memory pools.
```

Comment:
```
// Since we can reserve an extra amount of memory above (when taskBytes > bytes),
// ExceededMemoryLimitException can be thrown when reserving this extra amount.
// This will happen after "bytes" is moved from the operator context to the task context.
// At this point "bytes" isn't tracked by the operator context.
// When drivers get destroyed, only memory tracked in its operators are freed.
// Therefore, we have to cleanup the memory here.